core/src/main/java/io/questdb/cairo/CairoEngine.java

```diff
@@ -310,6 +318,108 @@ private void rename0(Path path, CharSequence tableName, Path otherPath, CharSequ
+
+        row.ts = clock.getTicks();
+        row.event = event;
+        telemetryPubSeq.done(cursor);
+    }
+
+    private class TelemetryWriterJob extends SynchronizedJob implements Closeable {
+        private final NanosecondClock nanosecondClock = configuration.getNanosecondClock();
+        private final MicrosecondClock microsecondClock = configuration.getMicrosecondClock();
+        private final CharSequence telemetryTableName = "telemetry";
+        private final QueueConsumer<TelemetryRow> myConsumer = this::toTelemetryTable;
+        private final CharSequence id;
+        private final TableWriter writer;
+
+        public TelemetryWriterJob(CairoConfiguration configuration) {
+            final CharSequence root = configuration.getRoot();
+
+            try(Path path = new Path()) {
+                if (getStatus(AllowAllCairoSecurityContext.INSTANCE, path, telemetryTableName) == TableUtils.TABLE_DOES_NOT_EXIST) {
+                    final TelemetryTableModel telemetry = new TelemetryTableModel(telemetryTableName);
+                    final AppendMemory appendMem = new AppendMemory();
+                    id = toGuid(nanosecondClock.getTicks(), microsecondClock.getTicks());
+
+                    telemetry.addColumn("ts", ColumnType.TIMESTAMP);
+                    telemetry.addColumn("id", ColumnType.STRING);
+                    telemetry.addColumn("event", ColumnType.SHORT);
+
+                    TableUtils.createTable(
+                            configuration.getFilesFacade(),
+                            appendMem,
+                            path,
+                            root,
+                            telemetry,
+                            configuration.getMkDirMode()
+                    );
+                } else {
+                    try (TableReader reader = new TableReader(configuration, telemetryTableName)) {
+                        final StringSink sink = new StringSink();
+                        reader.getCursor().getRecord().getStr(1, sink);
+                        id = sink;
+                    }
+                }
+            }
+
+            this.writer = new TableWriter(configuration, telemetryTableName);
+            addEvent(TelemetryEvent.UP);
+        }
+
+        private CharSequence toGuid(long mostSigBits, long leastSigBits) {
+            return (digits(mostSigBits >> 32, 8) + "-" + digits(mostSigBits >> 16, 4) + "-" + digits(mostSigBits, 4)
+                    + "-" + digits(leastSigBits >> 48, 4) + "-" + digits(leastSigBits, 12));
+        }
+
+        private CharSequence digits(long val, int digits) {
+            long hi = 1L << (digits * 4);
+            return Long.toHexString(hi | (val & (hi - 1))).substring(1);
+        }
+
+        private void addEvent(short event) {
+            final TelemetryRow row = new TelemetryRow();
+
+            row.ts = microsecondClock.getTicks();
+            row.id = id;
+            row.event = event;
+            toTelemetryTable(row);
```

Comment:
storing own events can be done via publishing to the queue, it will make handling of events more consistent and avoid allocating object instances to pass method args

Reply:
done